lib/FrontendTool/FrontendTool.cpp

```diff
@@ -453,9 +482,44 @@ void ABIDependencyEvaluator::reexposeImportedABI(
     addToABIExportMap(module, reexportedModule);
 }
 
+bool ABIDependencyEvaluator::isOverlayOfClangModule(ModuleDecl *swiftModule) {
+  assert(!swiftModule->isNonSwiftModule());
+
+  auto cacheEntry = isOverlayCache.find(swiftModule);
+  if (cacheEntry != isOverlayCache.end())
+    return cacheEntry->second;
+
+  llvm::SmallPtrSet<ModuleDecl *, 8> importList;
+  ::getImmediateImports(swiftModule, importList,
+                        {ModuleDecl::ImportFilterKind::Public});
+  bool isOverlay =
+      llvm::any_of(importList, [&](ModuleDecl *importedModule) -> bool {
+        return isClangOverlayOf(swiftModule, importedModule);
+      });
+  isOverlayCache[swiftModule] = isOverlay;
+  return isOverlay;
+}
+
+bool ABIDependencyEvaluator::isFakeCycleThroughOverlay(
+    ModuleDecl **sandwichModuleIter) {
```

Comment:
I might defensively assert that `sandwichModuleIter` is an element in `searchStack`, since that's what guarantees that all the dereferencing and pointer arithmetic is valid:

```cpp
  assert(sandwichModuleIter >= searchStack.begin() && sandwichModuleIter < searchStack.end()
         && "sandwichModuleIter points to an element in searchStack");
```

But that might just be me being paranoid.

Reply:
Paranoia is good. Little bit of arithmetic won't hurt anyone.